torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -629,6 +644,83 @@ void RequestCallbackImpl::processRpc(
       markComplete(std::move(CleanupAutogradContextResp()).toMessage());
       return;
     }
+    case MessageType::RUN_WITH_PROFILING_REQ: {
+      auto& rpcWithProfilingReq = static_cast<RpcWithProfilingReq&>(rpc);
+      auto wrappedMsgType = rpcWithProfilingReq.wrappedMessageType();
+      const auto profilingConfig = rpcWithProfilingReq.getProfilingConfig();
+      auto wrappedRpcResponseFuture = std::make_shared<FutureMessage>();
+      auto fromWorkerId = rpcWithProfilingReq.fromWorkerId();
+      // Enable the profiler with the config from the sender.
+      std::vector<torch::autograd::profiler::Event> profiledEvents;
+      {
+        torch::autograd::profiler::TLSProfilerGuard g(
+            profilingConfig,
+            [&profiledEvents](const std::vector<
+                              std::vector<torch::autograd::profiler::Event>>&
+                                  event_lists) {
+              // Gather all events into a vector
+              for (auto& l : event_lists) {
+                for (auto& e : l) {
+                  profiledEvents.push_back(e);
+                }
+              }
+              // find __start_profile event.
+              const torch::autograd::profiler::Event* profilerStart = nullptr;
+              for (const auto& e : profiledEvents) {
+                if (0 == strcmp(e.name(), "__start_profile")) {
+                  profilerStart = &e;
+                  break;
+                }
+              }
+              // We should always find __start_profile.
+              TORCH_CHECK(profilerStart != nullptr);
+              // For every event, precompute the relative CPU time before
+              // serialization.
+              for (const auto& e : profiledEvents) {
+                double cpu_elapsed_us = profilerStart->cpu_elapsed_us(e);
```

Comment:
@ilia-cher The basic logic here is,
1) similar to parse cpu trace, we find the start_profile event
2) for each event (whether it's a push or pop) we compute `cpu_elapsed_us` as would be done by python, but we do this on the remote end, so that we are using the correct start_profile event that has timestamp we need
3) we save this value and set it in the event. Then in deserialization, we set this on the event.
4) Python parse_cpu_trace checks if it is set. If so, we conclude that it is a remote event and use these preset values (instead of recomputing as is the normal python flow, which would be incorrect for the remote event case)